File: COWRUN.py
"""
---------------luyencode.net----------------------------------
---------------Problem: COWRUN-------------------------------
---------------Author: HeHe---------------------------------
---------------Time: 2024-10-19 17:08:38.755032+07:00---------
"""
in1 = lambda: int(input())
in2 = lambda: list(map(int, input().split()))
in3 = lambda: map(int, input().split())
MOD = 1000000007
# --------------- Solve ---------------  # 
n, m = in3()
D = [0] * n
for i in range(n):
    D[i] = in1()

dp = [[-1] * 2 for _ in range(n + 1)]
dp[n][m] = 0

for i in range(n - 1, -1, -1):
    if dp[i + 1][m - 1] != -1:
        dp[i][m] = max(D[i] + dp[i + 1][m - 1], dp[i + 1][m])
    else:
        dp[i][m] = dp[i + 1][m]

    for j in range(1, m):
        if dp[i + 1][j - 1] != -1 and dp[i + 1][j + 1] != -1:
            dp[i][j] = max(D[i] + dp[i + 1][j - 1], dp[i + 1][j + 1])
        elif dp[i + 1][j - 1] != -1:
            dp[i][j] = D[i] + dp[i + 1][j - 1]
        elif dp[i + 1][j + 1] != -1:
            dp[i][j] = dp[i + 1][j + 1]
    
    if dp[i + 1][1] != -1:
        dp[i][0] = dp[i + 1][1]

print(dp[0][m])


##########################################

File: COWRUN.cpp
/*
---------------luyencode.net----------------------------------
---------------Problem: COWRUN-------------------------------
---------------Author: HeHe---------------------------------
---------------Time: 2024-10-19 17:33:01.018875+07:00---------
*/
#include <iostream>
#include <iomanip>
#include <vector>
#include <queue>
#include <map>
#include <unordered_map>
#include <set>
#include <unordered_set>
#include <assert.h>
#include <algorithm>
#include <functional>
#include <numeric>
#include <math.h>
#include <cstdint>
#include <climits>

#define FAST_IO ios::sync_with_stdio(false);cin.tie(0);cout.tie(0);
using namespace std;

const int MOD = 1000000007;
int dp[10001][501] = {-1};

int main() {
#ifndef ONLINE_JUDGE
    freopen("input.txt", "r", stdin);
    freopen("output.txt", "w", stdout);
#endif
    FAST_IO;
    // --------------- Solve -------------------
    std::fill(&dp[0][0], &dp[0][0] + sizeof(dp) / sizeof(dp[0][0]), -1);
    int n, m;
    cin >> n >> m;
    vector<int> D(n);
    for (auto &x : D) { cin >> x; }
    dp[n][m] = 0;
    
    for (int i = n - 1; i >= 0; --i) {
        // case j == m
        if (dp[i + 1][m - 1] != -1) {
            dp[i][m] = max(D[i] + dp[i + 1][m - 1], dp[i + 1][m]);
        }
        else {
            dp[i][m] = dp[i + 1][m];
        }
        // case j (1 -> m - 1)
        for (int j = 1; j <= m - 1; ++j){
            int max_val = -1;
            // run
            if (dp[i + 1][j - 1] != -1) {
                max_val = max(max_val, D[i] + dp[i + 1][j - 1]);
            }
            // rest
            if (i + m - j <= n) {
                max_val = max(max_val, dp[i + m - j][m]);
            }
            dp[i][j] = max_val;
        }
        // case j == 0
        if (i + m <= n) {
            dp[i][0] = dp[i + m][m];
        }
    }

    cout << dp[0][m];
    ////////////////////////////////////////////
    return 0;
}
